flink-state-backends/flink-statebackend-forst/src/main/java/org/apache/flink/state/forst/fs/ByteBufferReadableFSDataInputStream.java

```diff
@@ -0,0 +1,195 @@
+            int inputStreamCapacity) {
+        this.originalInputStream = originalInputStream;
+        this.inputStreamBuilder = inputStreamBuilder;
+        this.readInputStreamPool = new LinkedBlockingQueue<>(inputStreamCapacity);
+    }
+
+    /**
+     * Reads up to <code>ByteBuffer#remaining</code> bytes of data from the input stream into a
+     * ByteBuffer. Not Thread-safe yet since the interface of sequential read of ForSt only be
+     * accessed by one thread at a time.
+     *
+     * @param bb the buffer into which the data is read.
+     * @return the total number of bytes read into the buffer.
+     * @exception IOException If the first byte cannot be read for any reason other than end of
+     *     file, or if the input stream has been closed, or if some other I/O error occurs.
+     * @exception NullPointerException If <code>bb</code> is <code>null</code>.
+     */
+    public int readFully(ByteBuffer bb) throws IOException {
+        if (bb == null) {
+            throw new NullPointerException();
+        } else if (bb.remaining() == 0) {
+            return 0;
+        }
+        return readFullyFromFSDataInputStream(originalInputStream, bb);
+    }
+
+    /**
+     * Reads up to <code>ByteBuffer#remaining</code> bytes of data from the specific position of the
+     * input stream into a ByteBuffer. Tread-safe since the interface of random read of ForSt may be
+     * concurrently accessed by multiple threads. TODO: Support to split this method to other class.
+     *
+     * @param position the start offset in input stream at which the data is read.
+     * @param bb the buffer into which the data is read.
+     * @return the total number of bytes read into the buffer.
+     * @exception IOException If the first byte cannot be read for any reason other than end of
+     *     file, or if the input stream has been closed, or if some other I/O error occurs.
+     * @exception NullPointerException If <code>bb</code> is <code>null</code>.
+     */
+    public int readFully(long position, ByteBuffer bb) throws Exception {
+        if (bb == null) {
+            throw new NullPointerException();
+        } else if (bb.remaining() == 0) {
+            return 0;
+        }
+
+        FSDataInputStream fsDataInputStream = readInputStreamPool.poll();
+        if (fsDataInputStream == null) {
+            fsDataInputStream = inputStreamBuilder.call();
+        }
+
+        int result;
+        if (fsDataInputStream instanceof PositionedReadable) {
+            byte[] tmp = new byte[bb.remaining()];
+            ((PositionedReadable) fsDataInputStream).readFully(position, tmp, 0, tmp.length);
+            bb.put(tmp);
+            result = tmp.length;
+        } else {
+            fsDataInputStream.seek(position);
+            result = readFullyFromFSDataInputStream(fsDataInputStream, bb);
+        }
+
+        if (!readInputStreamPool.offer(fsDataInputStream)) {
+            fsDataInputStream.close();
+        }
```

Comment:
I suggest putting the "readInputStreamPool.offer(fsDataInputStream)"  into a "try-finally" block, so that ensure all the inputStreams can be closed when an IO exception is encountered.